### PR TITLE
RGW - POSIX Driver - Reading sideloaded object works.

### PR DESCRIPTION
**Pull request by dang**
- Includes storing attributes as xattrs.
- Generate and write out etag if it doesn't exist.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/driver/posix/rgw_sal_posix.cc [hunk 1]
```diff
@@ -1178,39 +1184,145 @@ int POSIXObject::close(const DoutPrefixProvider* dpp)
...
+    return ret;
+
+  ret = source->get_obj_attrs(y, dpp);
+  if (ret < 0)
+    return ret;
+
+  auto iter = source->get_attrs().find(RGW_ATTR_ETAG);
+  if (iter == source->get_attrs().end()) {
+    /* Sideloaded file.  Generate necessary attributes. Only done once. */
+    int ret = source->generate_attrs(dpp, y);
```
Comment (cbodley) [inline src/rgw/driver/posix/rgw_sal_posix.cc:1250]: this seems like a bad idea in general, as the files could be arbitrarily large. if the S3 API requires us to return *some* value for ETag, is there a fake string we could return instead? either something like "NOT-A-REAL-ETAG", or if clients expect it to look like an MD5 hash, then return the hash of an empty buffer?

Comment (mattbenjamin) [reply src/rgw/driver/posix/rgw_sal_posix.cc:1250]: HTTP only requires some functional properties of etag--i.e., it's relationship to cache-control, I believe.  S3 I think actually requires it to be the MD5 hash of the byte stream in the object.  If it's the md5sum of an empty buffer, the client can complain about things, and I think boto3 at least does, for what it is worth.

**Comment by cbodley** [comment]
i spent some time experimenting with this etag header for GetObj specifically. i found that neither boto nor s3cmd tries to verify the checksum of the GetObj response body

when the etag is omitted: https://gist.github.com/cbodley/4497ec07853bbf90645328811766cf41

when the etag is faked: https://gist.github.com/cbodley/8ac10990a4c1c253b2cf627bcd313572

boto *does* complain if PutObj returns an ETag that's different than it calculated:
> boto.exception.S3DataError: BotoClientError: ETag from S3 did not match computed MD5. "" vs. b'37b51d194a7513e45b56f6524f2d51f2'

but the posix driver can handle PutObj etags just fine

**Comment by cbodley** [comment]
p.s. we also return ETags in ListObjects, but according to https://docs.aws.amazon.com/AmazonS3/latest/API/API_Object.html that ETag field of the ListBucketResult is optional

**Comment by mattbenjamin** [comment]
> p.s. we also return ETags in ListObjects, but according to https://docs.aws.amazon.com/AmazonS3/latest/API/API_Object.html that ETag field of the ListBucketResult is optional

Ok, fine with me.

File: src/rgw/driver/posix/rgw_sal_posix.cc (hunk 1 above)
Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:1250]: This is temporary, just to get s3-cmd to stop complaining.  In fact, much of this is temporary, as part of the PoC build.

Comment (dang) [reply src/rgw/driver/posix/rgw_sal_posix.cc:1250]: And also to test reading/writing xattrs

**Pull request merged**